standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -2364,91 +2364,129 @@ public Materialization getMaterializationInvalidationInfo(
+    StringBuilder queryUpdateDelete = new StringBuilder();
+    StringBuilder queryCompletedCompactions = new StringBuilder();
+    StringBuilder queryCompactionQueue = new StringBuilder();
+    // compose a query that select transactions containing an update...
+    queryUpdateDelete.append("SELECT \"CTC_UPDATE_DELETE\" FROM \"COMPLETED_TXN_COMPONENTS\" WHERE \"CTC_UPDATE_DELETE\" ='Y' AND (");
+    queryCompletedCompactions.append("SELECT 1 FROM \"COMPLETED_COMPACTIONS\" WHERE (");
+    queryCompactionQueue.append("SELECT 1 FROM \"COMPACTION_QUEUE\" WHERE (");
+    int i = 0;
+    for (String fullyQualifiedName : creationMetadata.getTablesUsed()) {
+      ValidWriteIdList tblValidWriteIdList =
+              validReaderWriteIdList.getTableValidWriteIdList(fullyQualifiedName);
+      if (tblValidWriteIdList == null) {
+        LOG.warn("ValidWriteIdList for table {} not present in creation metadata, this should not happen", fullyQualifiedName);
+        return null;
+      }
+
+      // First, we check whether the low watermark has moved for any of the tables.
+      // If it has, we return true, since it is not incrementally refreshable, e.g.,
+      // one of the commits that are not available may be an update/delete.
+      ValidWriteIdList currentTblValidWriteIdList =
+              currentValidReaderWriteIdList.getTableValidWriteIdList(fullyQualifiedName);
+      if (currentTblValidWriteIdList == null) {
+        LOG.warn("Current ValidWriteIdList for table {} not present in creation metadata, this should not happen", fullyQualifiedName);
+        return null;
+      }
+      if (!Objects.equals(currentTblValidWriteIdList.getMinOpenWriteId(), tblValidWriteIdList.getMinOpenWriteId())) {
+        LOG.debug("Minimum open write id do not match for table {}", fullyQualifiedName);
+        return null;
+      }
+
+      // ...for each of the tables that are part of the materialized view,
+      // where the transaction had to be committed after the materialization was created...
+      if (i != 0) {
+        queryUpdateDelete.append("OR");
+        queryCompletedCompactions.append("OR");
+        queryCompactionQueue.append("OR");
+      }
+      String[] names = TxnUtils.getDbTableName(fullyQualifiedName);
+      assert (names.length == 2);
+      queryUpdateDelete.append(" (\"CTC_DATABASE\"=? AND \"CTC_TABLE\"=?");
+      queryCompletedCompactions.append(" (\"CC_DATABASE\"=? AND \"CC_TABLE\"=?");
+      queryCompactionQueue.append(" (\"CQ_DATABASE\"=? AND \"CQ_TABLE\"=?");
+      params.add(names[0]);
+      params.add(names[1]);
+      queryUpdateDelete.append(" AND (\"CTC_WRITEID\" > " + tblValidWriteIdList.getHighWatermark());
+      queryCompletedCompactions.append(" AND (\"CC_HIGHEST_WRITE_ID\" > " + tblValidWriteIdList.getHighWatermark());
+      queryUpdateDelete.append(tblValidWriteIdList.getInvalidWriteIds().length == 0 ? ") " :
+              " OR \"CTC_WRITEID\" IN(" + StringUtils.join(",",
```

Comment:
This could be problematic in ORACLE where it fails when the IN clause contains more than 1000 elements. Are we okay with this for all of the queries?

Reply:
Thanks for point out this.
This can fail on Oracle if an MV source table is affected by more than 1000 open/aborted transaction when the MV was created (`tblValidWriteIdList.getInvalidWriteIds()`) or being rebuild (`currentValidTxnList.getInvalidTransactions()`).

Also found this pattern elsewhere: https://github.com/apache/hive/blob/c8d9d2c3b0bcf871145f49185344935b0c23da0e/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java#L357

A possible solution could be to split the list of invalid transaction ids to smaller lists
a) and run the assembled query with each sublist
or
b) assemble a query like:
```
SELECT CTC_UPDATE_DELETE FROM COMPLETED_TXN_COMPONENTS
WHERE 
...
CTC_WRITEID > ? OR (
  CTC_WRITEID IN (<sublist1>) OR
  CTC_WRITEID IN (<sublist2>) OR
  ...
  CTC_WRITEID IN (<sublistn>)
... 
```

Reply:
Is this 1000 limit in Oracle configurable? Wondering whether we could run into issues if it is. Also, not clear whether other RDBMSs may have a similar limit.
I guess we are using an IN clause because we assume its execution will be more performant for the RDBMSs backing HMS. However, it could be that those RDBMSs are already optimizing such queries internally. Variant with equals would always work. Should we consider that option?
If we decide to stick to the IN clause, I believe we could aim at emitting a single query to the RDBMS (iiuc that would be option b) and be rather conservative in the size of each sublist (e.g., 50 elements?).

Reply:
It seems that this 1000 limitation is not configurable.

Found that `TxnUtils.buildQueryWithINClauseStrings` is used elsewhere. This function can produce a list of queries not just based on the number of expressions in the `in` operator but the overall size of the direct sql query text. However we already have multiple in operators: at least two per source table. 

Since this issue is not introduced by this patch and all available solutions are complex I recommend dealing this in a follow-up patch: [HIVE-24925](https://issues.apache.org/jira/browse/HIVE-24925)

Reply:
LGTM